### PR TITLE
JDK-8329624: Add visitors for preview language features

### PR DESCRIPTION
**Pull request by jddarcy**
When new language features are added, the javax.lang.model may need to be updated. For certain classes of features, the API update includes introducing a new set of concrete visitors to handle the language feature.

The API scaffolding to support the new feature tends to be considerably larger than the API specifically for the new feature.

To aid work in progress (such as https://github.com/openjdk/jdk/pull/18509) and anticipated in the future, I think it would be helpful to introduce a persistent set of preview visitors independent of any particular language change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8329634](https://bugs.openjdk.org/browse/JDK-8329634) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8329624](https://bugs.openjdk.org/browse/JDK-8329624): Add visitors for preview language features (**Enhancement** - P4)
 * [JDK-8329634](https://bugs.openjdk.org/browse/JDK-8329634): Add visitors for preview language features (**CSR**)

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [45bbc292](https://git.openjdk.org/jdk/pull/18609/files/45bbc292b115b5be780ffbb3cef06e3cdc29b46c)
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**) ⚠️ Review applies to [45bbc292](https://git.openjdk.org/jdk/pull/18609/files/45bbc292b115b5be780ffbb3cef06e3cdc29b46c)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18609/head:pull/18609` \
`$ git checkout pull/18609`

Update a local copy of the PR: \
`$ git checkout pull/18609` \
`$ git pull https://git.openjdk.org/jdk.git pull/18609/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18609`

View PR using the GUI difftool: \
`$ git pr show -t 18609`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18609.diff">https://git.openjdk.org/jdk/pull/18609.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18609#issuecomment-2035513797)

### CODE REVIEW

**Comment by jddarcy** [comment]
> When new language features are added, the javax.lang.model may need to be updated. For certain classes of features, the API update includes introducing a new set of concrete visitors to handle the language feature.
> 
> The API scaffolding to support the new feature tends to be considerably larger than the API specifically for the new feature.
> 
> To aid work in progress (such as #18509) and anticipated in the future, I think it would be helpful to introduce a persistent set of preview visitors independent of any particular language change.

When one of the preview visitors is subclassed, javac emits a warning like:

```
javac VisitorTest.java 
VisitorTest.java:2: warning: [preview] ElementKindVisitorPreview is a reflective preview API and may be removed in a future release.
public class VisitorTest extends javax.lang.model.util.ElementKindVisitorPreview {
                                                      ^
1 warning
```

Once these preview visitors are present, the particular API to support language change P can be added and reviewed without being dwarfed in the overall scaffolding. The anticipated usage would be:

Methods particular to supporting preview feature P can be annotated accordingly to indicate that usage.
When P goes to a final feature, a new set of JDK $N visitors would be inserted between to latest numbered set of visitors and the preview visitors. Methods particular to P would be moved up into the new JDK $N visitors.

For example, if feature P when final in JDK 25, there would be a new set of JDK 25 concrete visitors extended the existing JDK 14 visitors and superclass of the preview visitors would be changed from the JDK 14 visitor to the corresponding JDK 25 visitor. Methods associated with feature P would be moved up from the preview visitors into the JDK 25 visitors.

File: src/java.compiler/share/classes/javax/lang/model/util/SimpleElementVisitorPreview.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2019, 2024, Oracle and/or its affiliates. All rights reserved.
```
Comment (vicente-romero-oracle) [inline src/java.compiler/share/classes/javax/lang/model/util/SimpleElementVisitorPreview.java:2]: general, please check the copyright years

**Review by vicente-romero-oracle — APPROVED** [review]
looks good

File: src/java.compiler/share/classes/javax/lang/model/util/SimpleElementVisitorPreview.java (hunk 1 above)
Comment (jddarcy) [reply src/java.compiler/share/classes/javax/lang/model/util/SimpleElementVisitorPreview.java:2]: Starting with 2019 was intention as a copied the corresponding FooVisitor14 file into FooVisitorPreview and then began editing.

Comment (vicente-romero-oracle) [reply src/java.compiler/share/classes/javax/lang/model/util/SimpleElementVisitorPreview.java:2]: ok

**Comment by AlanBateman** [comment]
The visitor classes are themselves preview APIs. Is this the first time that preview APIs have been proposed without a JEP? Just wondering how much visibility this preview API will get.

**Comment by jddarcy** [comment]
> The visitor classes are themselves preview APIs. Is this the first time that preview APIs have been proposed without a JEP? Just wondering how much visibility this preview API will get.

To clarify the intention, the visitor classes are meant to be persistent preview APIs to help host specific preview APIs associated with modeling upcoming new language features. In particular, these visitor classes are expected to get small updates as part of the work being done in

https://github.com/openjdk/jdk/pull/18509

The preview visitor classes as such don't need much visibility as their use will be through various previewing language features.

**Review by lahodaj — APPROVED** [review]
Looks good to me, thanks!

**Pull request closed without merging**